src/roman_datamodels/datamodels/_core.py

```diff
@@ -48,6 +47,22 @@ def wrapper(self, *args, **kwargs):
     return wrapper
 
 
+@contextmanager
+def _temporary_update_filename(datamodel, filename):
+    """
+    Context manager to temporarily update the filename of a datamodel so that it
+    can be saved with that new file name without changing the current model's filename
+    """
+    from roman_datamodels.stnode import Filename
+
+    if "meta" in datamodel._instance and "filename" in datamodel._instance.meta:
+        old_filename = datamodel._instance.meta.filename
+        datamodel._instance.meta.filename = Filename(filename)
+
+        yield
+        datamodel._instance.meta.filename = old_filename
```

Comment:
This specifically exists so that the `DataModel` instance being saved via `save` or `to_asdf` does not update its `filename`, the filename is only updated in the newly written ASDF file and the model which can be opened from that.

I chose to do it this way because `AsdfFile.write_to` is employed by these `to_asdf` (which is called by `save`), which writes the `AsdfFile` object to the new file descriptor passed to it, but it does not update the file descriptor of the `AsdfFile` object. This means the asdf file wrapped by the `DataModel` instance is not changing, so the filename of the python object being saved should not be updated.